bin/runtime-common/src/messages_batch_extension.rs

```diff
@@ -281,6 +314,61 @@ impl<Runtime, GrandpaInstance, ParachainsInstance, MessagesInstance> SignedExten
 	}
 }
 
+/// Extracts expected relay chain state from the call.
+fn extract_expected_relay_chain_state<Runtime, GrandpaInstance>(call: &CallOf<Runtime>) -> Option<ExpectedRelayChainState> where
+	Runtime: GrandpaConfig<GrandpaInstance>,
+	GrandpaInstance: 'static,
+	<Runtime as GrandpaConfig<GrandpaInstance>>::BridgedChain: Chain<BlockNumber = RelayBlockNumber>,
+	CallOf<Runtime>: IsSubType<CallableCallFor<GrandpaPallet<Runtime, GrandpaInstance>, Runtime>>,
+{
+	if let Some(GrandpaCall::<Runtime, GrandpaInstance>::submit_finality_proof {
+		ref finality_target,
+		..
+	}) = call.is_sub_type() {
+		return Some(ExpectedRelayChainState { best_block_number: *finality_target.number() });
+	}
+	None
+}
+
+/// Extracts expected parachain state from the call.
+fn extract_expected_parachain_state<Runtime, GrandpaInstance, ParachainsInstance>(call: &CallOf<Runtime>) -> Option<ExpectedParachainState> where
+	Runtime: GrandpaConfig<GrandpaInstance> + ParachainsConfig<ParachainsInstance, BridgesGrandpaPalletInstance = GrandpaInstance>,
+	GrandpaInstance: 'static,
+	ParachainsInstance: 'static,
+	<Runtime as GrandpaConfig<GrandpaInstance>>::BridgedChain: Chain<BlockNumber = RelayBlockNumber, Hash = RelayBlockHash, Hasher = RelayBlockHasher>,
+	CallOf<Runtime>: IsSubType<CallableCallFor<ParachainsPallet<Runtime, ParachainsInstance>, Runtime>>,
+{
+	// TODO: check para id (we'll need to bound instance of messages pallet with some ParaId)
```

Comment:
so we'll have a messages pallet instance for each Para we bridge, right? does this mean we have different `pallet_utility::Call::<Runtime>::batch_all` calls one for each ParaId?

Reply:
> so we'll have a messages pallet instance for each Para we bridge, right?

In theory we may serve different chains with the same instance of messages pallet, but it'll be too complex. So yes - 1 messages instance per 1 bridged chain. Please mind that for us this bridged chain is the other bridge hub, not the Statemin?. So we may serve XCM messages from any Polkadot/Kusama parachain using single messages pallet.

> does this mean we have different pallet_utility::Call::<Runtime>::batch_all calls one for each ParaId?

I'm not sure I got this part. In this extension we'll support 2 types of batch calls (relay+parachain+messages and parachain+messages). And we expect this parachain transaction to bring single parachain header - yes. If relayer wants to update the other parachain (which we don't plan to do at least initially) then yeah - he can craft the separate batch or submit a dedicated transaction.

Reply:
Ok, thanks, it's clear now. I misinterpreted/mismapped-in-my-brain `parahead` to be head of msg source parachain (Statemine), and that's why I was confused and thought about how will we transfer messages coming from other parachains :see_no_evil: 

But yes, that was a brainfart. `parahead` here is other BridgeHub and messages pallet instance is same regardless of message source (at most we might have different lanes later on, but still single pallet instance).